### PR TITLE
Open dynamically built notebook containing code and data in Iodide.

### PR DESCRIPTION
**Pull request by Octocat1337**
for now we can't automate the upload template file part.
Problem:
we can upload correctly but we cannot process the HTTP Response from iodide server.

### CODE REVIEW

**Comment by gmierz** [comment]
See the comments I sent to you over Riot. We can fully automate the POST request using a scripted form element. That doesn't solve the local iodide server requirement though.

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (gmierz) [inline .vscode/settings.json]: Add this directory to the `.gitignore`, we don't want this file in tree.

File: analyzer.py [hunk 2]
```diff
@@ -77,4 +85,68 @@ def ttest(self):
...
+
+    '''
+    
+    @app.route('/post',methods=['GET'])
+    def send_template():
+        post_to_iodide()
+    
+    '''
+
+    webbrowser.open('upload_file.html')
```
Comment (gmierz) [inline analyzer.py]: When I run the modified changes locally, there are two windows opened by this to the same page doing the same post request. Not sure why it happens though.

File: plotTest.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import numpy as np
```
Comment (gmierz) [inline plotTest.py]: You'll want to remove this file from the PR.

**Review by gmierz — CHANGES_REQUESTED** [review]
Great start here! We're almost there now! :)

File: analyzer.py (hunk 2 above)
Comment (Octocat1337) [reply analyzer.py]: fixed.
Due to Flask's debug mode, it is detecting changes on 2 processes.

**Review by gmierz — CHANGES_REQUESTED** [review]
You will have to rebase all of your changes onto the newest changes in the starter branch so we can see the Travis-CI build.

File: template_upload_file.html [hunk 4]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html>
+<body>
+	<script>
+		function post(path, params, method='post') {
+		  const form = document.createElement('form');
+		  form.method = method;
+		  form.action = path;
+		  form.id = 'uploadform';
+		  /*
```
Comment (gmierz) [inline template_upload_file.html]: Remove commented code.

File: template_upload_file.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		  form.submit();
+		}
+
+		// Template is read locally, and inserted here
+		// It will look like this after - an issue with this is that we will need to escape all `'`
+		// otherwise, this will results in javascript failures.
+		//var template = '%% md\n# Welcome to Iodide!\n\n\n<p id="htmlInMd"></p>\n\n\n%% js\n\ndocument.getElementById("htmlInMd").innerHTML = "parsedData[0].Address";\n\nvar data = document.getElementById("file0");\n\n\n\n%% fetch\n\n\n\n%% py \n\n\n\n'
+		var template = replace_me
+
+		// POST requests strip away new lines. Replace them here with a unique character pattern,
```
Comment (gmierz) [inline template_upload_file.html]: Remove commented code.

File: upload_file.html [hunk 6]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
```
Comment (gmierz) [inline upload_file.html]: This file shouldn't be added to the repo - it's dynamically generated.

File: analyzer.py [hunk 7]
```diff
@@ -77,4 +76,16 @@ def ttest(self):
                         'ts2': entry2['data']
                     })
 
-        return results
\ No newline at end of file
+        return results
+    '''
```
Comment (gmierz) [inline analyzer.py]: Remove all the code that was added here, we don't need it.

File: template_upload_file.html [hunk 8]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html>
+<body>
+	<script>
+		function post(path, params, method='post') {
+		  const form = document.createElement('form');
```
Comment (gmierz) [inline template_upload_file.html]: Inconsistent tabs/spaces used here. Make sure that we always use 4 spaces or 1 tab of 4 spaces.

File: perftestnotebook.py [hunk 9]
```diff
@@ -210,6 +216,41 @@ def main():
...
+
+    html = ''
+    with open('template_upload_file.html', 'r') as f:
+        html = f.read()
+
+    html = html.replace('replace_me', repr(tdata))
+    with open('upload_file.html', 'w+') as f:
+        f.write(html)
+    
+    webbrowser.open_new_tab('upload_file.html')
```
Comment (gmierz) [inline perftestnotebook.py]: Does this work when there is no browser open?

File: perftestnotebook.py [hunk 10]
```diff
@@ -184,7 +190,7 @@ def main():
                 new_results['ttest'].append(res)
 
         # Pretty print the results
```
Comment (gmierz) [inline perftestnotebook.py]: You can remove this `'analysis' in config` section, we don't need it anymore.

File: perftestnotebook.py (hunk 9 above)
Comment (Octocat1337) [reply perftestnotebook.py]: yes it will automatically open a new browser and a new tab in it.

File: perftestnotebook.py [hunk 11]
```diff
@@ -206,5 +212,40 @@ def main():
...
+    template = 'testing/template/template.txt'
+    tdata = ''
+    with open(template, 'r') as f:
+        tdata = f.read()
+
+    html = ''
+    with open('template_upload_file.html', 'r') as f:
+        html = f.read()
+
+    html = html.replace('replace_me', repr(tdata))
```
Comment (gmierz) [inline perftestnotebook.py]: All of this replacement and writing should be done above on line 153 in process.

File: perftestnotebook.py [hunk 12]
```diff
@@ -206,5 +212,40 @@ def main():
...
+    with open('upload_file.html', 'w+') as f:
+        f.write(html)
+    
+    webbrowser.open_new_tab('upload_file.html')
```
Comment (gmierz) [inline perftestnotebook.py]: Place this and everything below (down to app.run) into an `if 'analysis` in config:` condition so it only runs if there was an analysis was specified.

**Review by gmierz — CHANGES_REQUESTED** [review]

File: analyzer.py [hunk 13]
```diff
@@ -1,7 +1,5 @@
-import numpy as np
 import scipy.stats as stats
 
-
```
Comment (gmierz) [inline analyzer.py]: Leave this blank line.

File: analyzer.py [hunk 14]
```diff
@@ -1,7 +1,5 @@
-import numpy as np
```
Comment (gmierz) [inline analyzer.py:1]: Leave this import.

File: template_upload_file.html [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+                    }
+                }
+
+
+                document.body.appendChild(form);
+                form.submit();
+            }
+
+            // Template is read locally, and inserted here
+            // Need to escape all `'`
```
Comment (gmierz) [inline template_upload_file.html]: Add TODO to the start of this line. You should also make the sentence look a bit better, i.e.:
```
// Need to escape all `'`, otherwsie, this will result in
// javascript failures.
```

File: analyzer.py [hunk 16]
```diff
@@ -83,3 +83,12 @@ def ttest(self):
                     )
 
         return results
+
```
Comment (gmierz) [inline analyzer.py]: Remove all of this added code, it's not required.

File: notebookparser.py [hunk 17]
```diff
@@ -21,4 +21,12 @@ def parse_args():
     parser.add_argument(
         "--debug", action="store_true", default=False, help="Enable additional debug logging.",
     )
+    parser.add_argument(
+        "--no-iodide",
+        "-ni",
+        required=False,
```
Comment (gmierz) [inline notebookparser.py]: CLI arguments are optional by default, you can remove the required setting.

File: notebookparser.py [hunk 18]
```diff
@@ -21,4 +21,12 @@ def parse_args():
...
+        action="store_true",
+        default=False,
+        help="Start notebook without posting to iodide server"
```
Comment (gmierz) [inline notebookparser.py]: A better help message would be "Run this tool without starting the iodide server at the end."

File: perftestnotebook.py [hunk 19]
```diff
@@ -95,14 +102,18 @@ def parse_file_grouping(self, file_grouping):
...
         if not files:
-            raise Exception("Could not find any files in this configuration: %s" % file_grouping)
+            raise Exception(
+                "Could not find any files in this configuration: %s" % file_grouping
+            )
 
         return files
 
-    def process(self):
+    def process(self, args):
```
Comment (gmierz) [inline perftestnotebook.py]: Don't pass all args, you only need to pass `no_iodide` here.

File: perftestnotebook.py [hunk 20]
```diff
@@ -51,10 +54,14 @@ def __init__(self, file_groups, config, custom_transform=None):
                 obj = getattr(module, name)
                 if isinstance(obj, type) and issubclass(obj, tfmr.Transformer):
                     self.transformer = obj(files=[])
-                    logger.info("Found %s transformer" % self.transformer.__class__.__name__)
+                    logger.info(
```
Comment (gmierz) [inline perftestnotebook.py]: It seems like you might be using the wrong black formatting - can you make sure you run black but with a max line length of 100?

File: perftestnotebook.py [hunk 21]
```diff
@@ -144,66 +158,92 @@ def process(self):
...
+        Analyze data
+        """
         if "analysis" in self.config:
             # Analyze the data
-            all_results = {}
-            self.analyzer.data = fmt_data
-            for func in self.config["analysis"]:
-                all_results[func] = getattr(self.analyzer, func)()
-            return all_results
+            # TODO
```
Comment (gmierz) [inline perftestnotebook.py]: Undo this change, without this code, there's no way to run an analysis function.

The analyzer functions need to be changed to return code (in a string). Then, you need to take this returned code and insert it into a text file. This text file will then be uploaded as the "template file".

File: perftestnotebook.py [hunk 22]
```diff
@@ -144,66 +158,92 @@ def process(self):
...
+        else:
+            self.post_to_iodide(output_data_filepath)
 
+        return self.fmt_data
 
-def main():
-    args = parse_args()
+    def post_to_iodide(self, output_data_filepath):
+        # Upload template file to Iodide
+        template = "testing/template/template.txt"
```
Comment (gmierz) [inline perftestnotebook.py]: As I mentioned above, we shouldn't be displaying the template.txt file. We need to build a template.txt file and display that one.

I imagine that this would have two stages to start. It would start with inserting some preamble, like a generic title of the report as well as the fetch for the data, and the next stage would add the code to run in the notebook.

File: perftestnotebook.py [hunk 23]
```diff
@@ -144,66 +158,92 @@ def process(self):
...
+            # for analysis_method in analysis:
+            #   analyzer.process(analysis_method,self.fmt_data)
+            pass
 
-        return self.fmt_data
+        """
+        Post to iodide server
+        """
+        if args.no_iodide:
+            pass
```
Comment (gmierz) [inline perftestnotebook.py]: It would be better to have `if not no_iodide:` and not handle the other case.

File: testing/configs/config_artifactdownloader_output.yaml [hunk 24]
```diff
@@ -6,3 +6,6 @@ file_groups:
     path: testing/resources/artifactdownloader_output
     artifact: perfherder-data
     run_number: 1
+
+analysis:
```
Comment (gmierz) [inline testing/configs/config_artifactdownloader_output.yaml]: Please remove the analysis section from this config and make a new config if you need it.

File: testing/template/template.txt [hunk 25]
```diff
@@ -0,0 +1,30 @@
+%% md
```
Comment (gmierz) [inline testing/template/template.txt]: We won't need this file because you will create it dynamically in the analysis step.

**Review by gmierz — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+#-----Project specific-----
+
+# vscode
+.vscode
+
+# file to be uploaded to iodide
+upload_file.html 
+
+# test outputs
+testing/outputs/
```
Comment (gmierz) [inline .gitignore]: Add new line to end of file.

File: .travis.yml [hunk 27]
```diff
@@ -13,6 +13,8 @@ install:
   - "pip install -r requirements.txt"
 
 script:
-  - "python perftestnotebook.py --config=testing/configs/config_single_json_test.yaml"
+  - "python perftestnotebook.py --config=testing/configs/config_single_json_test.yaml -ni"
+  - "python perftestnotebook.py --config=testing/configs/config_analyzer.yaml -ni"
   - "python -m black -l 100 ."
   - "python -m pytest -v testing/tests/"
+
```
Comment (gmierz) [inline .travis.yml]: Remove whitespace on this line.

File: testing/resources/template/header [hunk 28]
```diff
@@ -0,0 +1,18 @@
...
+
+%% fetch
+
+text: data_string = http://127.0.0.1:5000/data
+
+%% js
+
+var data_object = JSON.parse(data_string);
+
+
```
Comment (gmierz) [inline testing/resources/template/header]: You don't need to have anything below this line.

File: testing/resources/analysis-templates/scatterplot [hunk 29]
```diff
@@ -0,0 +1,13 @@
+import matplotlib.pyplot as plt
```
Comment (gmierz) [inline testing/resources/analysis-templates/scatterplot]: Each analysis should be self-contained. So you should have this import here also: `from js import data_object`

Also, start the section with a `%% py`.

File: testing/resources/analysis-templates/scatterplot [hunk 30]
```diff
@@ -0,0 +1,13 @@
...
+
+
+fig,ax = plt.subplots()
```
Comment (gmierz) [inline testing/resources/analysis-templates/scatterplot]: You can do `plt.figure()` here, no need for the axis or figure object.

File: testing/resources/analysis-templates/scatterplot [hunk 31]
```diff
@@ -0,0 +1,13 @@
...
+
+for element in data_object:
+    data_array = element["data"]
+    x = [x["xaxis"] for x in data_array]
+    y = [x["value"] for x in data_array]
+    label = element["name"]+"\n"+element["subtest"]
+    ax.scatter(x,y,label=label)
```
Comment (gmierz) [inline testing/resources/analysis-templates/scatterplot]: `plt.scatter(x, y, label=label)`

File: testing/resources/analysis-templates/scatterplot [hunk 32]
```diff
@@ -0,0 +1,13 @@
...
+plt.legend()
+plt.show()
```
Comment (gmierz) [inline testing/resources/analysis-templates/scatterplot]: Add a blank line at the end.

File: perftestnotebook.py [hunk 33]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+        """
+        Analyze data
+        """
+        analyzer_templates = ""
         if "analysis" in self.config:
-            # Analyze the data
-            all_results = {}
-            self.analyzer.data = fmt_data
+            
+            print(self.config["analysis"])
```
Comment (gmierz) [inline perftestnotebook.py]: Remove all print statements like this.

File: perftestnotebook.py [hunk 34]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+            # without readLine
             for func in self.config["analysis"]:
-                all_results[func] = getattr(self.analyzer, func)()
-            return all_results
+                analyzer_templates = analyzer_templates + self.analyzer.get_analysis_template(func)
```
Comment (gmierz) [inline perftestnotebook.py]: You can do `analyzer_templtes += self.analyzer.get_notebook_section`

File: perftestnotebook.py [hunk 35]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+
+        print("Writing results to %s" % output_data_filepath)
+
+        with open(output_data_filepath, "w") as f:
+            json.dump(self.fmt_data, f, indent=4, sort_keys=True)
+
+        """
+        Analyze data
+        """
+        analyzer_templates = ""
```
Comment (gmierz) [inline perftestnotebook.py]: Rename this to `notebook`

File: analyzer.py [hunk 36]
```diff
@@ -33,53 +33,30 @@ def split_subtests(self):
...
-                for entry2 in split_data[subtest]:
-                    if entry2["name"] == name:
-                        continue
-                    if (
-                        "%s-%s" % (name, entry2["name"]) in done
-                        or "%s-%s" % (entry2["name"], name) in done
-                    ):
-                        continue
-                    done["%s-%s" % (name, entry2["name"])] = True
+def main():
```
Comment (gmierz) [inline analyzer.py]: Make sure you remove all of this `main` stuff when you're closer to being ready for landing.

File: analyzer.py [hunk 37]
```diff
@@ -33,53 +33,30 @@ def split_subtests(self):
...
-        results = []
+    def get_header(self):
+        template_header_path = "testing/resources/template/header"
+        with open(template_header_path, "r") as f:
+            template_header_content = f.read()
+            return template_header_content
 
-        split_data = self.split_subtests()
+    def get_analysis_template(self,func):
+        template_function_folder_path = "testing/resources/analysis-templates/"
```
Comment (gmierz) [inline analyzer.py]: This is a constant value so it should be placed at the top of the file within a variable like `NOTEBOOK_SECTIONS_PATH`

File: analyzer.py [hunk 38]
```diff
@@ -33,53 +33,30 @@ def split_subtests(self):
...
-        """
-        results = []
+    def get_header(self):
+        template_header_path = "testing/resources/template/header"
+        with open(template_header_path, "r") as f:
+            template_header_content = f.read()
+            return template_header_content
 
-        split_data = self.split_subtests()
+    def get_analysis_template(self,func):
```
Comment (gmierz) [inline analyzer.py]: Make sure you run black on this.

Also, you will need to add a docstring comment for this method that describes what it does and what the inputs and outputs are. There are examples of this elsewhere in the code.

File: testing/resources/template/header [hunk 39]
```diff
@@ -0,0 +1,18 @@
+%% md
+# Welcome to PerftestNotebook
+
+press the :fast_forward: button on your top left corner to run whole notebook
```
Comment (gmierz) [inline testing/resources/template/header]: Please use capitals and periods in your writing.

File: analyzer.py [hunk 40]
```diff
@@ -33,53 +33,30 @@ def split_subtests(self):
...
+        template_function_folder_path = "testing/resources/analysis-templates/"
+        template_function_file_path = template_function_folder_path + func
```
Comment (gmierz) [inline analyzer.py]: You should use `os.path.join` here instead of the `+`

File: perftestnotebook.py [hunk 41]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+        """
+        Post to iodide server
+        """
+        if not no_iodide:
+           self.post_to_iodide(output_data_filepath,analyzer_templates)
 
         return self.fmt_data
 
+    def post_to_iodide(self, output_data_filepath,analyzer_templates):
+        """# Upload template file to Iodide"""
```
Comment (gmierz) [inline perftestnotebook.py]: Better docstring comment needed here.

File: perftestnotebook.py [hunk 42]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+        print(output_data_filepath)
+        template_header_path = "testing/resources/template/header"
```
Comment (gmierz) [inline perftestnotebook.py]: Constant value should be declared at the top of the script.

File: perftestnotebook.py [hunk 43]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+        template_content = ""
 
-def main():
-    args = parse_args()
+        with open(template_header_path, "r") as f:
+            template_content = template_content + f.read()
```
Comment (gmierz) [inline perftestnotebook.py]: You can do `template_content = f.read()` here, and below (it can be moved outside the `with` and changed to `template_content += notebook_sections`

File: perftestnotebook.py [hunk 44]
```diff
@@ -144,66 +158,94 @@ def process(self):
...
+            print(self.config["analysis"])
+            # without readLine
             for func in self.config["analysis"]:
-                all_results[func] = getattr(self.analyzer, func)()
-            return all_results
+                analyzer_templates = analyzer_templates + self.analyzer.get_analysis_template(func)
+            
+
+        """
+        Post to iodide server
```
Comment (gmierz) [inline perftestnotebook.py]: Change these multi-line comments to single line ones: `# Post to Iodide server`

**Review by gmierz — CHANGES_REQUESTED** [review]
A larger review this time since it's getting closer. I think the `analysis_templates` names should be changed to `notebook_sections` and we should also use the term `notebook_section` when we refer to a single section instead of the `templates` naming pattern (except for the template HTML file, since that is an actual template) you've been using. It makes what you are referring to clearer since we aren't really opening and playing with templates.

File: testing/resources/template/header (hunk 28 above)
Comment (Ralts1337) [reply testing/resources/template/header]: ok I see what you mean. but we still need the "from js import data_object" to pass that data from js to python.

File: analyzer.py [hunk 45]
```diff
@@ -33,53 +35,29 @@ def split_subtests(self):
...
+    def get_header(self):
+        template_header_path = "testing/resources/template/header"
+        with open(template_header_path, "r") as f:
+            template_header_content = f.read()
+            return template_header_content
+
+    def get_notebook_section(self, func):
+        template_function_folder_path = "testing/resources/notebook-sections/"
+        template_function_file_path = os.path.join(template_function_folder_path, func)
+        print(template_function_file_path)
```
Comment (gmierz) [inline analyzer.py]: Remove this print.

File: perftestnotebook.py [hunk 46]
```diff
@@ -144,66 +148,76 @@ def process(self):
...
+
+        # Gather config["analysis"] corresponding notebook sections
         if "analysis" in self.config:
-            # Analyze the data
-            all_results = {}
-            self.analyzer.data = fmt_data
             for func in self.config["analysis"]:
-                all_results[func] = getattr(self.analyzer, func)()
-            return all_results
+                notebook_sections = notebook_sections + self.analyzer.get_notebook_section(func)
```
Comment (gmierz) [inline perftestnotebook.py]: Use `notebook_sections += self.analyzer...`

File: template_upload_file.html [hunk 47]
```diff
@@ -0,0 +1,34 @@
...
+<html>
+    <body>
+        <script>
+            function post(path, params, method='post') {
+                const form = document.createElement('form');
+                form.method = method;
+                form.action = path;
+                form.id = 'uploadform';
+
+                for(const key in params){
```
Comment (gmierz) [inline template_upload_file.html]: Add spaces around the `(...)`

File: template_upload_file.html [hunk 48]
```diff
@@ -0,0 +1,34 @@
...
+                    if(params.hasOwnProperty(key)){
```
Comment (gmierz) [inline template_upload_file.html]: Same thing here.

File: template_upload_file.html [hunk 49]
```diff
@@ -0,0 +1,34 @@
...
+                        const textarea = document.createElement('textarea');
+                        textarea.name=key;
```
Comment (gmierz) [inline template_upload_file.html]: Spaces around the `=`

File: template_upload_file.html [hunk 50]
```diff
@@ -0,0 +1,34 @@
...
+                        textarea.value = params[key];
+                        textarea.style.display="none";
```
Comment (gmierz) [inline template_upload_file.html]: Spaces around the `=`

File: perftestnotebook.py [hunk 51]
```diff
@@ -1,14 +1,17 @@
 import json
 import os
 import pathlib
+import webbrowser
+
```
Comment (gmierz) [inline perftestnotebook.py:5]: You can remove this extra blank line

**Review by gmierz — CHANGES_REQUESTED** [review]
Excellent! Please resolve these final comments and then we can land it (don't forget about the analyzer.py main function removal).

File: perftestnotebook.py [hunk 52]
```diff
@@ -102,14 +105,15 @@ def parse_file_grouping(self, file_grouping):
 
         return files
 
-    def process(self):
+    def process(self, no_iodide):
```
Comment (gmierz) [inline perftestnotebook.py]: Sorry, one last thing to change is make the no_iodide an optional argument `no_iodide=True`.

**Review by gmierz — CHANGES_REQUESTED** [review]

**Comment by gmierz** [comment]
Wonderful - thanks for this work! Landing it.

**Review by gmierz — APPROVED** [review]
